torchgeo/datasets/utils.py

```diff
@@ -737,3 +739,15 @@ def percentile_normalization(
         (img - lower_percentile) / (upper_percentile - lower_percentile + 1e-5), 0, 1
     )
     return img_normalized
+
+
+def check_instance_type(objects: Union[Path, Iterable[Path]]) -> bool:
```

Comment:
Why is this needed?

Reply:
This function help to check for parameterized generic types such as ``PathLike[bytes]`` , the original isinstance fails to check parameterized generic types. Here is the the error ``Second argument to "isinstance" must be a class or tuple of classes TypeVar or generic type with type arguments not allowed``.

Reply:
It doesn't have a docstring, but I'm guessing this function returns True if the input is a pathlib.Path or an iterable of pathlib.Path? But then you always cast to str, which doesn't make sense for iterables. Also, `str(object)` will always be an instance of str, so that check doesn't make sense. Can't we just check if it's an instance of `pathlib.Path`?

Reply:
if we try to check if the parameter is an instance of pathlib.Path using ``isinstance`` method , Here are the results.

 Code
``isinstance(self.paths, Path)``
Errors:
```plaintext
torchgeo/datasets/geo.py:297: error: Parameterized generics cannot be used with class or instance checks  [misc]
torchgeo/datasets/geo.py:297: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; 
expected "_ClassInfo"  [arg-type]
```

The cast to str is required to find out if the path's type is str or bytes. Here is an example 
```from os import PathLike
from pathlib import Path
dir_path = Path("/home")
print(isinstance(dir_path,str))  # This will return false 
print(hasattr(dir_path,'decode'))
```

As you have mentioned str(object) will be an instance of str , so that check does not make sense and will incorrectly convert objects to string. Here is what i mean.

```from pathlib import Path
dir_path = Path.cwd()  # return type is <class 'pathlib.PosixPath'>
dir_path = Path("/home")
print(isinstance(str(dir_path), str))  # returns true
```

A better way of checking for strings would be using the ``hasattr(str_to_check, "encode")``.

Reply:
I'm guessing that error is because you're using:
```
from pathlib import Path
```
AND
```
from .utils import Path
```
and the names conflict. Not sure if we should use a different name or just use `pathlib.Path`.

Do we need to check whether or not it's bytes? Can we treat str and bytes identically?

Reply:
No, sir both of them have been tested in seperate files. `from pathlib import Path` and `from .utils import Path` are not imported simultaneously.  We can skip the test if desired so.

Reply:
I have the following file:
```python
from pathlib import Path

p = Path('.')

if isinstance(p, Path):
    print("Path")
else:
    print("str or bytes")
```
`mypy --strict` passes on this file. Anything I'm missing here?

I don't think you need to check `Iterable[Path]`, just `Path`.

Reply:
It throws the same error when i try to run the below code.
```import os
from typing import Union
from typing_extensions import TypeAlias

Path: TypeAlias = Union[str, bytes, os.PathLike[bytes], os.PathLike[str]]

testPath: Path = os.getcwd()

print(isinstance(testPath, Path))```

Reply:
Yes, because `torchgeo.datasets.utils.Path` and `pathlib.Path` are completely different things.